### PR TITLE
cache server

### PR DESCRIPTION
**Pull request by khibino**
cache server prototyping

- [x] quit process for forked threads
- [x] server implementation
- [x] server top-level

### CODE REVIEW

File: src/DNSC/Concurrent.hs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+module DNSC.Concurrent (
+  forkProcessQ,
+  forkLoop,
+  forksProcessQ,
+  forksLoop,
+  )where
```
Comment (kazu-yamamoto) [inline src/DNSC/Concurrent.hs:6]: should put a space char after `)`.

File: src/DNSC/Concurrent.hs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  let loop = do
+        isQuit <- readIORef qref
+        unless isQuit $ body *> loop
+  waitQuit <- forksWithWait $ replicate n loop
+  return $ writeIORef qref True *> waitQuit
+
+forksWithWait :: [IO ()] -> IO (IO ())
+forksWithWait bodies = do
+  waitQ <- newChan
+  sequence_ [ forkIO $ body *> writeChan waitQ () | body <- bodies ]
```
Comment (kazu-yamamoto) [inline src/DNSC/Concurrent.hs:43]: To manages threads, `async` must be used to prevent thread-leak.

File: src/DNSC/Server.hs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+type NE a = (a, [a])
+
+type Request s a = (s, (DNSHeader, NE Question), a)
+type Response s a = ((s, DNSMessage), a)
+
+udpSockets :: PortNumber -> [HostName] -> IO [(Socket, SockAddr)]
+udpSockets port = mapM aiSocket . filter ((== Datagram) . addrSocketType) <=< addrInfo port
+  where
+    aiSocket ai = (,) <$> S.socket (addrFamily ai) (addrSocketType ai) (addrProtocol ai) <*> pure (addrAddress ai)
```
Comment (kazu-yamamoto) [inline src/DNSC/Server.hs:32]: You can use `openSocket`.

File: src/DNSC/Concurrent.hs (hunk 1 above)
Comment (khibino) [reply src/DNSC/Concurrent.hs]: https://github.com/khibino/dns-resolver/pull/4/commits/3ccbb0d7c80c74abf80d237c75dca3aed134953a

File: src/DNSC/Concurrent.hs (hunk 2 above)
Comment (khibino) [reply src/DNSC/Concurrent.hs]: https://github.com/khibino/dns-resolver/pull/4/commits/81115aa4c89cc096f2d3c0eb70119d223b92c3f9

File: dns-resolver.cabal [hunk 4]
```diff
@@ -47,6 +52,14 @@ library
   default-language:    Haskell2010
   ghc-options:         -Wall
 
+Executable cache-server
+  main-is:             server.hs
+  build-depends:       base
+                     , dns-resolver
+  hs-source-dirs:      mains
+  default-language:    Haskell2010
+  ghc-options:         -Wall -threaded -with-rtsopts=-N2 -with-rtsopts=-qn1
```
Comment (kazu-yamamoto) [inline dns-resolver.cabal:61]: This should be `-Wall -threaded -rtsopts` as users want to specify the number of cores by themselves.

File: mains/server.hs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+      return opt { bindHosts = hosts }
+  where
+    (ars, hosts, errs) = getOpt RequireOrder descs args
+    helpOnLeft e = putStrLn e *> help *> return Nothing
+
+run :: ServerOptions -> IO ()
+run opts = Server.run (logLevel opts) (disableV6NS opts) (concurrency opts) (fromIntegral $ port opts) (bindHosts opts)
+
+main :: IO ()
+main = maybe (return ()) run =<< parseOptions =<< getArgs
```
Comment (kazu-yamamoto) [inline mains/server.hs:72]: Why did you choose left-to-right style? I feel that right-to-left style is more natural.

File: dns-resolver.cabal (hunk 4 above)
Comment (khibino) [reply dns-resolver.cabal:61]: This is just the default option, so the user can still specify any options they wish.

File: mains/server.hs (hunk 5 above)
Comment (khibino) [reply mains/server.hs:72]: The function composition operator `. ` are often mixed, I have basically used the right-to-left style.

Should I try to use the left-to-right style basically in this codebase?

File: src/DNSC/Concurrent.hs (hunk 2 above)
Comment (khibino) [reply src/DNSC/Concurrent.hs]: https://github.com/khibino/dns-resolver/pull/4/commits/ca3585fd34105cf76eeb9712de00245ca8ebfb1a

**Review by kazu-yamamoto — APPROVED** [review]
LTGM

File: src/DNSC/Concurrent.hs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  ) where
+
+import Control.Concurrent.Chan (newChan, readChan, writeChan)
+import Control.Monad (unless, replicateM_)
+import Data.IORef (newIORef, readIORef, writeIORef)
+
+import Control.Concurrent.Async (async, wait)
+
+
+forkProcessQ :: (a -> IO ())
```
[the remark this answers is not in the document]
Comment (khibino) [reply src/DNSC/Concurrent.hs]: https://github.com/khibino/dns-resolver/pull/4/commits/bf5d09badf06010f7b1d0e864a1d819f41753a2b

**Pull request merged**